### ISSUE TITLE
Feedback: 'Custom Serilog configuration'

### ISSUE THREAD

**Issue by dasjestyr** [Particular/docs.particular.net#2609]
Feedback for 'Custom Serilog configuration' https://docs.particular.net/samples/logging/serilog-custom/

Location in GitHub: https://github.com/Particular/docs.particular.net/blob/master/samples/logging/serilog-custom/sample.md

ping @SimonCropp

Is there a dependency on serilog already being initialized before .Use<SerilogFactory>() is called? I ask because my logging is setup in a separate windsor installer. I assumed it would all be handled via dependency injection so there shouldn't be a problem, but once I updated the LogManager, I noticed that NSB logs are no longer being output to the output window

**Comment by mikeminutillo** [comment]
Hi @dasjestyr,

When you first reference the `SerilogFactory` if grabs a copy of the `Log.Logger` and sets that as the default. If you later change the `Log.Logger`, the `SerilogFactory` needs to be informed of the new logger.

The recommended approach would be to configure your logging as early in the application lifecycle as possible. Otherwise you run the risk that important log messages that happen early are going to a different and unexpected location.

I don't know if @SimonCropp would consider this a bug but I raised it in the NServiceBus.Serilog repo here https://github.com/SimonCropp/NServiceBus.Serilog/issues/9 with a fix at https://github.com/SimonCropp/NServiceBus.Serilog/pull/10

**Comment by SimonCropp** [comment]
@dasjestyr i pushed a new nuget with the fix from @mikeminutillo. can you let us know if that fixes your problem

**Comment by SimonCropp** [comment]
@dasjestyr will close this for now. please feel free to re-open, or open a new issue, if the problem is not resolved.

**Comment by dasjestyr** [comment]
Sorry, this got lost in the mess. My logger is set up during Castle Windsor's installer cycle, as is NSB, so I'm not sure how that would have affected it. My logging installer actually registers an instance of my ILogger interface with an implementation, and then sets it on a service locator so that it can be called statically. It also registers serilog's ILogger. Like this:

``` csharp
public class LoggingInstaller : IWindsorInstaller
{
    public void Install(IWindsorContainer container, IConfigurationStore store)
    {
        container.Register(Component
            .For<Serilog.ILogger>()
            .Instance(GetSerilogger())
            .LifestyleSingleton());

        container.Register(Component
            .For<ILogger>()
            .ImplementedBy<SerilogLogger>()
            .LifestyleSingleton());

        Logger.SetLogger(HostContainer.Current.Resolve<ILogger>());
    }

    private static Serilog.ILogger GetSerilogger()
    {
        var logger = new LoggerConfiguration()
                .WriteTo.RollingFile(@"C:\Logs\DAS\DAS.Services.Surveys\log-{Date}.log",
                    retainedFileCountLimit: 5,
                    fileSizeLimitBytes: 10485760,
                    outputTemplate: "[{Level}]\t{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz}\t{Message}{Exception}{NewLine}")
                .WriteTo.Trace()
                .WriteTo.Loggly()
                .CreateLogger();

        return logger;
    }
}
```

NSB is also initialized in a different IWindsorInstaller, and that's where I set the logger on the LogManager. So, I don't actually know was LogManager.Use<SerilogFactory>() is doing under the hood or how it would jive with my DI. Is there a way I can set the implementation of Serilog.ILogger on LogManager?

**Comment by dasjestyr** [comment]
I'm going to try this, but I'm not sure if that is lazy loaded so i may not work

``` csharp
public class NsbInstaller : IWindsorInstaller
{
    public void Install(IWindsorContainer container, IConfigurationStore store)
    {
        string licensePath = $"{AppDomain.CurrentDomain.BaseDirectory}/App_Data/License.xml";

        var logFactory = LogManager.Use<SerilogFactory>();
        logFactory.WithLogger(container.Resolve<Serilog.ILogger>());

        var config = new EndpointConfiguration(ServiceEndpoints.SurveysWrite);
```

**Comment by SimonCropp** [comment]
@dasjestyr since we need the logger to be configured before ant static classes of NSB are used, you need to perform the logging setup very early in you app lifecycle. https://docs.particular.net/nservicebus/logging/#when-to-configure-logging

So the NSB logging is not compatible with dependency injection

**Comment by dasjestyr** [comment]
Oof, that's no good. I wonder if there's a way (or could be a way) to plug it into NSB's DI pipeline. I mean, it already allows you to attach your own container, so it doesn't seem far off to get logging in there.

**Comment by SimonCropp** [comment]
you can place your own logger instance in the container and have that injected for your own handlers and sagas etc. the reason NSB does not use this approach for its own logging is that the majority of logging frameworks are not optimized for quickly spinning up a logger instance.

So for example if you inject a logger into a handler with the same lyfecycle, then every time a new instance of the handler is created then a new logger is instantiated. instatiating a logger is generally a very expensive option and can have a significant impact on performance.

So with that in mind all NSB classes have a static field that is the type specific logger. Hence we need logging configured prior to NSB classes being used. 

Also in NSB we want to perform logging during the phases where DI is being configured and used at app startup. So again it prevents is from using DI to configure logging.

You can still take your DI approach for your own classes, although you should probably measure the perf impact, but you also need to do a once off call to `LogManager.Use` at app startup

**Comment by dasjestyr** [comment]
Ok, I'll have to just do it after the container finishes initializing and lose some of the NSB startup logging :(

I could just set up serilog and set it on NSB and register the instance with the container before starting NSB, from within the NSB installer... but... ugh what a mess. I just wanted to keep things in their own installers for maintenance reasons. I could probably move NSB out of the installer into its own thing that occurs after container initialization is complete...

On a side note, I actually have the logger instance registered with windsor as a singleton, though so I don't think repeated instantiation would be a problem, or at least, that's how I thought it would operate. Is that not right?

**Comment by SimonCropp** [comment]
sorry. yes singleton in the container should mean you have no perf impact for your own classes

>  but... ugh what a mess. I just wanted to keep things in their own installers for maintenance reasons

I understand that but from a framework perspective it is very difficult to make this work in a way that is intuitive for the consumer and behave in a way that has predictable perf.

> Ok, I'll have to just do it after the container finishes initializing and lose some of the NSB startup logging 

You need to be careful with this. you dont just "lose some of the NSB startup logging" you lose the logging for any NSB type that has been used prior to logging being configured, ie any time that type is used in the future it will use the default NSB logging.

**Comment by dasjestyr** [comment]
Ok, so what I've done is I've removed the IWindsorInstaller interface from the NsbInstaller that I was talking about. Now, I initialize windsor, and then I pass the container to the NsbInstaller and use it to resolve ILogger and set LogManager with it, then I configure and start the endpoint. I'm not getting any errors and it seems that I'm getting all of the NSB logs (at least in the startup process) sent to loggly

**Comment by SimonCropp** [comment]
good to hear. sorry for the friction

**Comment by dasjestyr** [comment]
Hey no, problem and thanks for talking it through with me! I think this will work just fine. The nsb startup logic is still contained within it's own "installer" that is simply not auto-scanned and run during container initialization. A perfect compromise methinks. :)
